src/sidechain.cpp

```diff
@@ -0,0 +1,466 @@
+// Copyright (c) 2017-2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <sidechain.h>
+
+#include <arith_uint256.h>
+#include <coins.h>
+#include <consensus/validation.h>
+#include <hash.h>
+#include <primitives/transaction.h>
+#include <streams.h>
+#include <undo.h>
+#include <util/check.h>
+
+#include <algorithm>
+#include <cstdint>
+
+void CreateDBUndoData(CTxUndo &txundo, const uint8_t type, const COutPoint& record_id, const Coin& encoded_data) {
```

Comment:
do we also need to unwind sidechain db operations (like incrementing of `DBIDX_SIDECHAIN_WITHDRAW_PROPOSAL_ACKS`) in the undo data?

Reply:
ah, I see now that `CreateDbEntry` and `DeleteDbEntry` create undo data, and `ModifyDbEntry` calls both of them. Nevermind. Leaving this comment here for future reviewers